### ISSUE TITLE
[Feature Request] Optional Epoch Size arg for Engine.run()

### ISSUE THREAD

**Issue by DNGros** [pytorch/ignite#103]
Thanks for your work on the library. It contains a lot of helpful functions to get clean out boilerplate code.

I have had some difficulty using ignite with infinite data iterators (in this case coming from Torchtext), as the run will never end or fire epoch end events. It would be nice if Engine.run() implementations like Trainer and Evaluator took in an optional batches_per_epoch argument. This would be assigned on the state object. Engine._run_once_on_dataset() could check if this attribute was not None, and in that case only go through the next State.batches_per_epoch batches. Otherwise, the functionality would be the same.

**Comment by jasonkriss** [comment]
@DNGros Seems pretty reasonable. I can't say I have used many infinite data iterators myself. Can you point me in the direction of some specific examples so I can familiarize myself?

**Comment by DNGros** [comment]
@jasonkriss 
For example in https://github.com/pytorch/text/blob/master/torchtext/data/iterator.py#L13 the __iter__() function will infinitely produce batches by default. Of course as I went to look at this code to link to it I noticed that it provides a "repeat" arg that when set to false makes my issue with the ignite Trainer go away.

Sorry, I should have noticed this. I think part of my initial confusion by the interface came from still getting used to Pytorch after having being more used to Keras which has a fit_generator() function that takes in a similar steps_per_epoch parameter. It makes more sense for this to be needed when passing in a generator rather than an iterator, so unless their is also desire for the Trainer class to support generators, I see now that this feature is not as needed unless others have issues with it.
